### DISCUSSION TITLE
In streaming mode, how would you add final reponse to chat history?

### DISCUSSION THREAD

**Discussion by longpk** [microsoft/semantic-kernel#8105] [Q&A] [answered]
I'm having this chatbot with function calling in streaming mode as follows. I don't know how to correctly add the final responese to the chat history. 

    userInput = None
    while True:
        userInput = input("User > ")

        if userInput == "exit":
            break

        history.add_user_message(userInput)

        response = chat_completion.get_streaming_chat_message_contents(
            chat_history=history,
            settings=execution_settings,
            kernel=kernel,
            arguments=KernelArguments(),
        )

        result = None
        print("Assistant > ", end="")
        async for chunk in response:
            print(str(chunk[0]), end="")
            if result is None:
                result = chunk[0]
            else:
                result = result + chunk[0]
        print("")

        # Don't know how to add response to history. The following line would raise exception in the next chat turn.
        history.add_message(result)

The code will cause this exception:

    semantic_kernel.exceptions.service_exceptions.ServiceResponseException: ("<class 'semantic_kernel.connectors.ai.open_ai.services.azure_chat_completion.AzureChatCompletion'> service failed to complete the prompt", BadRequestError('Error code: 400 - {\'error\': {\'message\': "An assistant message with \'tool_calls\' must be followed by tool messages responding to each \'tool_call_id\'. The following tool_call_ids did not have response messages: call_pAvHKESwoWEMDDyNfitPu0Ja, call_0V4KTml2pfM1sPPsYIsvg9N0", \'type\': \'invalid_request_error\', \'param\': \'messages.[9].role\', \'code\': None}}'))

**Comment by akordowski** [comment]
In C# there is an `Metadata` property on the `StreamingChatMessageContent` class. The last message contains `Metadata["FinishReason"] = "stop"` with which you can check if the message is the last. I hope you can find something similar in Python.

**Comment by moonbox3** [comment, marked as answer]
Hi @longpk, in Python, we have the following auto function calling concept sample script that shows how to handle streaming messages, and adding them to the chat history when done. You can see in the `async def handle_streaming(...)` method, that we build up the `result_content` list, if applicable, and then join those items once the method is done.

https://github.com/microsoft/semantic-kernel/blob/9d4966ab777d5bb124d64dc349ffcc1d37b7dc66/python/samples/concepts/auto_function_calling/chat_gpt_api_function_calling.py#L120

**Reply by longpk** [reply]
Thank you for pointing to the right direction. Looking at sample script, I changed this line:
`history.add_message(result)`
to 
`history.add_assistant_message(result)`

and it works. 
Thanks for your help.
